README.md
# Volto Image Block (by kitconcept)

[![NPM](https://img.shields.io/npm/v/@kitconcept/volto-image-block.svg)](https://www.npmjs.com/package/@kitconcept/volto-image-block)
[![Build Status](https://github.com/kitconcept/volto-image-block/actions/workflows/code.yml/badge.svg)](https://github.com/kitconcept/volto-image-block/actions)
[![Build Status](https://github.com/kitconcept/volto-image-block/actions/workflows/unit.yml/badge.svg)](https://github.com/kitconcept/volto-image-block/actions)
[![Build Status](https://github.com/kitconcept/volto-image-block/actions/workflows/acceptance.yml/badge.svg)](https://github.com/kitconcept/volto-image-block/actions)

![kitconcept GmbH](https://raw.githubusercontent.com/kitconcept/volto-form-builder/master/kitconcept.png)

## Introduction

Custom implementation for the image component used in the image block. Few customizations in the edit and view mode and in the sidebar schema.

It features srcsets and blurhash.

## Compatibility

|Version   |Volto version |
|----------|--------------|
|<1.0.0    |>=16.x.x      |
|>=1.0.0   |<=17.0.0a0    |


CHANGELOG.md
# kitconcept's volto-image-block Release Notes

<!-- You should *NOT* be adding new change log entries to this file.
     You should create a file in the news directory instead.
     For helpful instructions, please see:
     https://6.dev-docs.plone.org/volto/developer-guidelines/contributing.html#create-a-pull-request
-->

<!-- towncrier release notes start -->

## 1.0.3 (2023-07-18)

### Bugfix

- Support block data with external image url @reebalazs [#18](https://github.com/kitconcept/volto-export/pull/18)


## 1.0.2 (2023-07-14)

### Bugfix

- Support block data with missing image_scales, after image upload @reebalazs [#16](https://github.com/kitconcept/volto-export/pull/16)
- Fix upload image preview in edit mode @sneridagh [#17](https://github.com/kitconcept/volto-export/pull/17)


## 1.0.1 (2023-06-08)

### Bugfix

- Fix hydration warning in console. @sneridagh [#11](https://github.com/kitconcept/volto-export/pull/11)
- Remove project-specific code in Caption component. @davisagli [#13](https://github.com/kitconcept/volto-export/pull/13)


## 1.0.0 (2023-03-15)

### Breaking

- Suport for Volto 17, Node 18 and fastblurhash working in webpack5. Support for Webpack 5 means this can only work in Volto >=17 @sneridagh [#10](https://github.com/kitconcept/volto-export/pull/10)


## 0.2.0 (2023-03-14)

### Feature

- Add towncrier for changelog @sneridagh [#4](https://github.com/kitconcept/volto-export/pull/4)
- Better image loader options for Volto defaults for source sets [#5](https://github.com/kitconcept/volto-export/pull/5)
- Use dockerized add-on development @sneridagh [#9](https://github.com/kitconcept/volto-export/pull/9)

### Bugfix

- Move dataAdapter to the block config so it can also be overriden in nested blocks via blocksConfig @sneridagh [#7](https://github.com/kitconcept/volto-export/pull/7)


acceptance/cypress/tests/basic.cy.js
import {
  getSlateEditorAndType,
  getSelectedSlateEditor,
} from '../support/slate';

context('Basic Acceptance Tests', () => {
  describe('Text Block Tests', () => {
    beforeEach(() => {
      // given a logged in editor and a page in edit mode
      cy.autologin();
      cy.createContent({
        contentType: 'Document',
        contentId: 'document',
        contentTitle: 'Document',
      });
      cy.visit('/');
    });

    it('As editor I can add a page with a text block', function () {
      // when I add a page with a text block
      cy.get('#toolbar-add').click();
      cy.get('#toolbar-add-document').click();
      cy.get('.documentFirstHeading')
        .type('My Page')
        .get('.documentFirstHeading')
        .contains('My Page');

      getSlateEditorAndType(
        '.block .slate-editor [contenteditable=true]',
        'This is the text',
      );

      getSelectedSlateEditor().contains('This is the text');
      cy.get('#toolbar-save').click();
      cy.url().should('eq', Cypress.config().baseUrl + '/my-page');
    });

    it('As editor I can add a link to a text block', function () {
      cy.intercept('GET', `/**/*?expand*`).as('content');
      cy.visit('/document');
      cy.wait('@content');

      cy.navigate('/document/edit');

      // when I create a link
      getSlateEditorAndType(
        '.slate-editor [contenteditable=true]',
        'Colorless green ideas sleep furiously.',
      ).setSelection('furiously');

      cy.clickSlateButton('Add link');

      cy.get('.link-form-container input').type('https://google.com{enter}');
      cy.get('#toolbar-save').click();
      cy.wait('@content');
      cy.url().should('eq', Cypress.config().baseUrl + '/document');

      // then the page view should contain a link
      cy.get('.ui.container p').contains(
        'Colorless green ideas sleep furiously.',
      );
      cy.get('.ui.container p a')
        .should('have.attr', 'href')
        .and('include', 'https://google.com');
    });
  });
});


src/index.js
import { Img } from './components/ImageLoader';
import { imageBlockSchemaEnhancer } from './components/Blocks/Image/schema';
import { ImageBlockDataAdapter } from './components/Blocks/Image/adapter';
import ImageWidget from './components/ImageWidget/ImageWidget';
import FastBlurHashLoader from './components/FastBlurHashLoader/FastBlurHashLoader';

const applyConfig = (config) => {
  config.blocks.blocksConfig.image = {
    ...config.blocks.blocksConfig.image,
    schemaEnhancer: imageBlockSchemaEnhancer,
    dataAdapter: ImageBlockDataAdapter,
  };

  config.registerComponent({
    name: 'Image',
    component: Img,
  });

  config.widgets.widget.image = ImageWidget;

  config.settings.appExtras = [
    ...config.settings.appExtras,
    {
      match: '',
      component: FastBlurHashLoader,
    },
  ];

  return config;
};

export default applyConfig;


src/components/Blocks/Image/adapter.js
import { flattenScales } from '../../../helpers/flattenScalesURLs';

export const ImageBlockDataAdapter = ({
  block,
  data,
  id,
  item,
  onChangeBlock,
  value,
}) => {
  const SIZEMAP = {
    l: 'large',
    m: 'medium',
    s: 'small',
  };

  let dataSaved = {
    ...data,
    [id]: value,
  };

  if (id === 'align' && !(value === 'left' || value === 'right')) {
    if (data.size !== 'l') {
      dataSaved = {
        ...dataSaved,
        size: 'l',
        styles: {
          ...dataSaved.styles,
          'size:noprefix': SIZEMAP['l'],
        },
      };
    }
  }

  if (id === 'size') {
    dataSaved = {
      ...dataSaved,
      styles: {
        ...dataSaved.styles,
        'size:noprefix': SIZEMAP[value],
      },
    };
  }

  if (id === 'url') {
    const image = flattenScales(item);

    dataSaved = {
      ...dataSaved,
      credit: { data: item?.credit },
      description: item?.Description,
      title: item?.Title,
      image_scales:
        item?.image_scales || {
          ...(item?.image && { image: [image] }),
        } ||
        {},
    };
  }
  onChangeBlock(block, dataSaved);
};


src/components/ImageLoader/fast-blurhash.js
import { decodeBlurHash } from 'fast-blurhash';

// Note this code depends on the markup created by AnyLoader, so
// if it is changed, the drill-down to the image has to be adjusted here.

document.querySelectorAll('img.blurhash').forEach((img) => {
  const dataRaw = img.getAttribute('data');
  const data = JSON.parse(dataRaw);
  const { width, height, hash, canvasStyle, imgWidth, ratio } = data;
  // Take the original aspect ratio from the img
  const image = img.previousSibling.firstChild.firstChild;
  const computedStyle = getComputedStyle(image);
  const { aspectRatio, objectFit } = computedStyle;
  // Replace the image with a canvas
  //
  const canvas = document.createElement('canvas');
  img.replaceWith(canvas);
  for (const cssProp in canvasStyle) {
    canvas.style[cssProp] = canvasStyle[cssProp];
  }
  canvas.style.width = parseFloat(imgWidth) + 'px';
  canvas.height = height;
  canvas.width = width;
  // Important: ignore auto aspect ratios from the image,
  // ie. "auto 1440 / 980"
  if (
    aspectRatio &&
    !aspectRatio.startsWith('auto') &&
    !canvas.style.aspectRatio
  ) {
    canvas.style.aspectRatio = aspectRatio;
    canvas.style.objectFit = objectFit;
    canvas.style.height = 'auto';
  } else if (imgWidth) {
    canvas.style.height = parseFloat(imgWidth) / ratio + 'px';
  }
  // Paint the blurhash on the canvas
  const pixels = decodeBlurHash(hash, width, height);
  const ctx = canvas.getContext('2d');
  const imageData = ctx.createImageData(width, height);
  imageData.data.set(pixels);
  ctx.putImageData(imageData, 0, 0);
});


src/components/ImageLoader/webpack.fast-blurhash.config.js
var UglifyJsPlugin = require('uglifyjs-webpack-plugin');

var config = {
  entry: __dirname + '/fast-blurhash.js',
  output: {
    path: __dirname + '/dist',
    filename: 'fast-blurhash.min.js',
  },
  module: {
    rules: [
      {
        test: /\.jsx$/,
        loader: 'babel',
        exclude: /node_modules/,
        query: {
          presets: ['env', 'react'],
          plugins: ['transform-class-properties'],
        },
      },
      { test: /\.css$/, loaders: ['style', 'css'] },
    ],
  },
  devtool: 'eval-source-map',
};

/*
 * If bundling for production, optimize output
 */
if (process.env.NODE_ENV === 'production') {
  config.devtool = false;
  config.optimization = {
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        uglifyOptions: {
          compress: true,
          ecma: 6,
          mangle: true,
        },
        sourceMap: false,
      }),
    ],
  };
}

module.exports = config;


src/components/ImageLoader/test-data.js
export const blockDataSample1 = {
  '@type': 'image',
  align: 'center',
  credit: {},
  id: 'ad43f059-fab9-4b7b-b49f-d4ec1fb6e3fc',
  image_field: 'image',
  image_scales: {
    image: [
      {
        'content-type': 'image/jpeg',
        download: '@@images/image-1126-7bb80db0a452739fa96a97d3c6517495.jpeg',
        filename: 'image-light.jpg',
        height: 633,
        scales: {
          icon: {
            download: '@@images/image-32-d5575b5e0795407436104a16e18ebd41.jpeg',
            height: 17,
            width: 32,
          },
          large: {
            download:
              '@@images/image-800-9c8a61153197fca732c86aa1f2091cb3.jpeg',
            height: 449,
            width: 800,
          },
          larger: {
            download:
              '@@images/image-1000-36afc44f7c991c35735f3338e76bc0a7.jpeg',
            height: 562,
            width: 1000,
          },
          mini: {
            download:
              '@@images/image-200-6f2dd5c028b4fbf2c975d98f42077678.jpeg',
            height: 112,
            width: 200,
          },
          preview: {
            download:
              '@@images/image-400-4abbad67754c69a7af7a1b91d78e680e.jpeg',
            height: 224,
            width: 400,
          },
          teaser: {
            download:
              '@@images/image-600-705867124c8790a0fc7d4c799e506c2e.jpeg',
            height: 337,
            width: 600,
          },
          thumb: {
            download:
              '@@images/image-128-eebe3707668d4cd3752626c026191ca0.jpeg',
            height: 71,
            width: 128,
          },
          tile: {
            download: '@@images/image-64-0ea29ecf0ed3720e590d849dcbda5834.jpeg',
            height: 35,
            width: 64,
          },
        },
        size: 475285,
        width: 1126,
      },
    ],
  },
  size: 'l',
  title: 'Headline H2 ',
  url: '/example/image-2',
};

export const blockDataSample2 = {
  '@type': 'image',
  align: 'center',
  credit: {},
  id: 'ad43f059-fab9-4b7b-b49f-d4ec1fb6e3fc',
  image_field: 'image',
  image_scales: {
    image: [
      {
        'content-type': 'image/jpeg',
        download: '@@images/image-1126-7bb80db0a452739fa96a97d3c6517495.jpeg',
        filename: 'image-light.jpg',
        height: 633,
        scales: {
          icon: {
            download: '@@images/image-32-d5575b5e0795407436104a16e18ebd41.jpeg',
            height: 17,
            width: 32,
          },
          large: {
            download:
              '@@images/image-800-9c8a61153197fca732c86aa1f2091cb3.jpeg',
            height: 449,
            width: 800,
          },
          larger: {
            download:
              '@@images/image-1000-36afc44f7c991c35735f3338e76bc0a7.jpeg',
            height: 562,
            width: 1000,
          },
          mini: {
            download:
              '@@images/image-200-6f2dd5c028b4fbf2c975d98f42077678.jpeg',
            height: 112,
            width: 200,
          },
          preview: {
            download:
              '@@images/image-400-4abbad67754c69a7af7a1b91d78e680e.jpeg',
            height: 224,
            width: 400,
          },
          teaser: {
            download:
              '@@images/image-600-705867124c8790a0fc7d4c799e506c2e.jpeg',
            height: 337,
            width: 600,
          },
          thumb: {
            download:
              '@@images/image-128-eebe3707668d4cd3752626c026191ca0.jpeg',
            height: 71,
            width: 128,
          },
          tile: {
            download: '@@images/image-64-0ea29ecf0ed3720e590d849dcbda5834.jpeg',
            height: 35,
            width: 64,
          },
        },
        size: 475285,
        width: 1126,
      },
    ],
  },
  size: 'l',
  title: 'Headline H2 ',
  // an url that MUST be flattened, and we test for it
  url: 'http://localhost:3000/example/image-2',
};

export const blockDataSample3 = {
  // Teaser from object browser has @id instead of url
  '@id': 'http://localhost:3000/example/teaser-block/image',
  '@type': 'Document',
  Description:
    'Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.',
  Title: 'Headline H2',
  effective: '2022-12-15T13:28:23+00:00',
  getObjSize: '929.3 KB',
  getRemoteUrl: null,
  hasPreviewImage: true,
  head_title: null,
  image_field: 'preview_image',
  image_scales: {
    preview_image: [
      {
        'content-type': 'image/jpeg',
        download:
          '@@images/preview_image-1126-1a01db87b603934db947bb1d72a06ed8.jpeg',
        filename: 'bfs-image-2.jpg',
        height: 1126,
        scales: {
          icon: {
            download:
              '@@images/preview_image-32-55bf93d1a340ed478f81be2833a71993.jpeg',
            height: 32,
            width: 32,
          },
          large: {
            download:
              '@@images/preview_image-800-eb3aebd2b90ad5b03fcbdcef0be0beb5.jpeg',
            height: 800,
            width: 800,
          },
          larger: {
            download:
              '@@images/preview_image-1000-aa0b59eca1a9b588993187796d29e491.jpeg',
            height: 1000,
            width: 1000,
          },
          mini: {
            download:
              '@@images/preview_image-200-87345416346beb71801c3e0e6274125c.jpeg',
            height: 200,
            width: 200,
          },
          preview: {
            download:
              '@@images/preview_image-400-2df9d934007070e332dcb58d0d42be2f.jpeg',
            height: 400,
            width: 400,
          },
          teaser: {
            download:
              '@@images/preview_image-600-84157b40fce66bd541a0876f5534c125.jpeg',
            height: 600,
            width: 600,
          },
          thumb: {
            download:
              '@@images/preview_image-128-b0e482c9def76f77ef514904bc0a567b.jpeg',
            height: 128,
            width: 128,
          },
          tile: {
            download:
              '@@images/preview_image-64-0c3b9155a1112aaaf093e722f926fcb3.jpeg',
            height: 64,
            width: 64,
          },
        },
        size: 951578,
        width: 1126,
      },
    ],
  },
  mime_type: 'text/plain',
  title: 'Headline H2',
};

export const blockDataSample4 = {
  '@type': 'image',
  align: 'center',
  credit: {},
  id: 'ad43f059-fab9-4b7b-b49f-d4ec1fb6e3fc',
  // missing image_field
  image_scales: {
    image: [
      {
        'content-type': 'image/jpeg',
        download: '@@images/image-1126-7bb80db0a452739fa96a97d3c6517495.jpeg',
        filename: 'image-light.jpg',
        height: 633,
        scales: {
          icon: {
            download: '@@images/image-32-d5575b5e0795407436104a16e18ebd41.jpeg',
            height: 17,
            width: 32,
          },
          large: {
            download:
              '@@images/image-800-9c8a61153197fca732c86aa1f2091cb3.jpeg',
            height: 449,
            width: 800,
          },
          larger: {
            download:
              '@@images/image-1000-36afc44f7c991c35735f3338e76bc0a7.jpeg',
            height: 562,
            width: 1000,
          },
          mini: {
            download:
              '@@images/image-200-6f2dd5c028b4fbf2c975d98f42077678.jpeg',
            height: 112,
            width: 200,
          },
          preview: {
            download:
              '@@images/image-400-4abbad67754c69a7af7a1b91d78e680e.jpeg',
            height: 224,
            width: 400,
          },
          teaser: {
            download:
              '@@images/image-600-705867124c8790a0fc7d4c799e506c2e.jpeg',
            height: 337,
            width: 600,
          },
          thumb: {
            download:
              '@@images/image-128-eebe3707668d4cd3752626c026191ca0.jpeg',
            height: 71,
            width: 128,
          },
          tile: {
            download: '@@images/image-64-0ea29ecf0ed3720e590d849dcbda5834.jpeg',
            height: 35,
            width: 64,
          },
        },
        size: 475285,
        width: 1126,
      },
    ],
  },
  size: 'l',
  title: 'Headline H2 ',
  url: '/example/image-2',
};

export const blockDataNoScalesSample1 = {
  '@type': 'image',
  url: 'http://localhost:3000/image1234.jpg',
  // empty image_scales, missing image_field
  image_scales: {},
  align: 'center',
  size: 'l',
  styles: { 'size:noprefix': 'large' },
};

export const blockDataNoScalesSample2 = {
  '@type': 'image',
  // Teaser from object browser has @id instead of url
  '@id': 'http://localhost:3000/image1234.jpg',
  // empty image_scales, missing image_field
  image_scales: {},
  align: 'center',
  size: 'l',
  styles: { 'size:noprefix': 'large' },
};

export const blockDataNoScalesSample3 = {
  '@type': 'image',
  // External link
  url: 'https://placekitten.com/408/287',
  // empty image_scales, missing image_field
  image_scales: {},
  align: 'center',
  size: 'l',
  styles: { 'size:noprefix': 'large' },
};

export const blockDataNoScalesSample4 = {
  '@type': 'image',
  // External link
  url: 'https://kitconcept.com/logo.svg',
  // empty image_scales, missing image_field
  image_scales: {},
  align: 'center',
  size: 'l',
  styles: { 'size:noprefix': 'large' },
};

export const contentDataSample1 = {
  'content-type': 'image/png',
  download:
    'http://localhost:3000/images/plone-foundation.png/@@images/image-2000-9d895f5dd2600938c123550e14965f08.png',
  filename: null,
  height: 439,
  scales: {
    great: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-1200-79d8298ba97bbcac8957f2313a4f8739.png',
      height: 263,
      width: 1200,
    },
    huge: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-1600-36844277b711713c66c7c77f6eafa2bf.png',
      height: 351,
      width: 1600,
    },
    icon: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-32-471419cba465f6bf008e71a2a6597a40.png',
      height: 7,
      width: 32,
    },
    large: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-800-00b242dbee2491cdd98180ee7d0a4fe8.png',
      height: 175,
      width: 800,
    },
    larger: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-1000-39bec9ae81852eede0f930c9aced869b.png',
      height: 219,
      width: 1000,
    },
    mini: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-200-010a606b0d76b6a6e089e8ba466dedca.png',
      height: 43,
      width: 200,
    },
    preview: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-400-1766720795b7f683a5d9c9672b9fb259.png',
      height: 87,
      width: 400,
    },
    teaser: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-600-85dc73d6512c05dad7f2ad0bab8c256d.png',
      height: 131,
      width: 600,
    },
    thumb: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-128-b60c381eeb9c02111b3b8d5b82e4fcd8.png',
      height: 28,
      width: 128,
    },
    tile: {
      download:
        'http://localhost:3000/images/plone-foundation.png/@@images/image-64-b45e38d323633ae920af93e7e7e24656.png',
      height: 14,
      width: 64,
    },
  },
  size: 50737,
  width: 2000,
};

export const stringSample1 = 'http://localhost:3000/@@images/image/huge';

export const stringSample2 = 'https://placekitten.com/408/287';

export const stringSample3 = 'https://kitconcept.com/logo.svg';
